[FIX] dms: fix wrong test

yajo opened the pull request:
In v13, this test is programmed in such a way that the demo user is supposed to be able to copy that subdirectory: https://github.com/OCA/dms/blob/c3f802db43362127e70d8c7b4987fb71d4c1f01c/dms/tests/test_directory.py#L40

However, in https://github.com/OCA/dms/pull/7 that test was modified indicating that demo user didn't have permissions to do that: https://github.com/OCA/dms/blob/e3b6d8d24534f2a68bfb88e310cc70cefe46bb64/dms/tests/test_directory.py#L39

Rolling back that change to ensure premissions remain the same in both versions of the module.


In our downstream CI, that test (v12) is failing (which actually means that the test is passing when it was expected to fail) with these logs:

<details>

```
2020-09-23 09:05:38,705 40 INFO prod odoo.addons.dms.tests.common: Test (test_copy_sub_directory) with user (6) failed! 
2020-09-23 09:05:38,706 40 ERROR prod odoo.addons.dms.tests.test_directory: ERROR 
...
2020-09-23 09:05:40,957 40 INFO prod odoo.addons.dms.tests.test_directory: ====================================================================== 
2020-09-23 09:05:40,957 40 ERROR prod odoo.addons.dms.tests.test_directory: ERROR: test_copy_sub_directory (odoo.addons.dms.tests.test_directory.DirectoryTestCase) 
2020-09-23 09:05:40,957 40 ERROR prod odoo.addons.dms.tests.test_directory: Traceback (most recent call last): 
2020-09-23 09:05:40,957 40 ERROR prod odoo.addons.dms.tests.test_directory: `   File "/opt/odoo/auto/addons/dms/tests/common.py", line 83, in wrapper 
2020-09-23 09:05:40,957 40 ERROR prod odoo.addons.dms.tests.test_directory: `     raise Exception(_("Error has not been raised")) 
2020-09-23 09:05:40,957 40 ERROR prod odoo.addons.dms.tests.test_directory: ` Exception: Error has not been raised 
2020-09-23 09:05:40,957 40 INFO prod odoo.addons.dms.tests.test_directory: Ran 18 tests in 3.973s 
2020-09-23 09:05:40,957 40 ERROR prod odoo.addons.dms.tests.test_directory: FAILED 
2020-09-23 09:05:40,957 40 INFO prod odoo.addons.dms.tests.test_directory:  (errors=1) 
2020-09-23 09:05:40,957 40 ERROR prod odoo.modules.module: Module dms: 0 failures, 1 errors 
...
2020-09-23 09:05:42,329 40 INFO prod odoo.addons.dms.tests.common: Test (test_copy_sub_directory) with user (6) failed! 
2020-09-23 09:05:42,335 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: ERROR 
...
2020-09-23 09:05:46,501 40 INFO prod odoo.addons.dms.tests.common: Test (test_copy_sub_directory) with user (6) failed! 
2020-09-23 09:05:46,502 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: ERROR 
...
2020-09-23 09:05:48,867 40 INFO prod odoo.addons.dms.tests.test_directory_mail: ====================================================================== 
2020-09-23 09:05:48,867 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: ERROR: test_copy_sub_directory (odoo.addons.dms.tests.test_directory_mail.DirectoryActionTestCase) 
2020-09-23 09:05:48,867 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: Traceback (most recent call last): 
2020-09-23 09:05:48,867 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: `   File "/opt/odoo/auto/addons/dms/tests/common.py", line 83, in wrapper 
2020-09-23 09:05:48,867 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: `     raise Exception(_("Error has not been raised")) 
2020-09-23 09:05:48,867 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: ` Exception: Error has not been raised 
2020-09-23 09:05:48,867 40 INFO prod odoo.addons.dms.tests.test_directory_mail: ====================================================================== 
2020-09-23 09:05:48,868 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: ERROR: test_copy_sub_directory (odoo.addons.dms.tests.test_directory.DirectoryTestCase) 
2020-09-23 09:05:48,868 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: Traceback (most recent call last): 
2020-09-23 09:05:48,868 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: `   File "/opt/odoo/auto/addons/dms/tests/common.py", line 83, in wrapper 
2020-09-23 09:05:48,868 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: `     raise Exception(_("Error has not been raised")) 
2020-09-23 09:05:48,868 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: ` Exception: Error has not been raised 
2020-09-23 09:05:48,868 40 INFO prod odoo.addons.dms.tests.test_directory_mail: Ran 38 tests in 7.909s 
2020-09-23 09:05:48,868 40 ERROR prod odoo.addons.dms.tests.test_directory_mail: FAILED 
2020-09-23 09:05:48,868 40 INFO prod odoo.addons.dms.tests.test_directory_mail:  (errors=2) 
2020-09-23 09:05:48,868 40 INFO prod odoo.modules.module: odoo.addons.dms.tests.test_directory_mail tested in 7.91s, 15111 queries 
2020-09-23 09:05:48,868 40 ERROR prod odoo.modules.module: Module dms: 0 failures, 2 errors 
```

</details>

@Tecnativa TT25645

pedrobaeza commented:
@etobella what is the reason for changing this?

victoralmau commented:
Maybe, by default `demo` value in this function is `False` https://github.com/OCA/dms/blob/c3f802db43362127e70d8c7b4987fb71d4c1f01c/dms/tests/common.py#L177

yajo commented:
@pedrobaeza means why was this changed when backporting from v13.

etobella commented:
I remember changing it because it was failing on travis, but I couldn't find a reason why it was failing. Maybe I thought that I had changed the value on 13 and I needed to change it on 12.0. Let me review why it isn't working :disappointed:

etobella commented:
I found the difference, SVG files can only be uploaded if you are system, however, there is a difference between 12 and 13:

on 12, is system is defined by user, you must be from group administration https://github.com/odoo/odoo/blob/12.0/odoo/fields.py#L1831 https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/models/res_users.py#L756

on 13, is system is defined by environment, https://github.com/odoo/odoo/blob/13.0/odoo/fields.py#L1903, but in order to be system, your user must be administrator or come from a su environment (https://github.com/odoo/odoo/blob/eb534d078dcdf58aedf09131dd57c8f6da69a910/odoo/api.py#L515). The su environment is defined by the initial user, that must be superadmin, on tests, we are allways on superadmin, so, that should be allways True (https://github.com/odoo/odoo/blob/eb534d078dcdf58aedf09131dd57c8f6da69a910/odoo/api.py#L432)
If we change the folder, and it does not contain SVG files, That will probably work on a similar way on both versions.

etobella commented:
However, it is weird, as it works fine on local with the last version :dizzy_face:

etobella commented:
Well, after investigating it a little, I can only assume that travis is recognizing some filed on the folder as a svg, however, they are csv, ods, xls, xlsx. Maybe we can replace `self.directory_sub_demo_01 = self.browse_ref("dms.directory_03_demo")` by `self.directory_sub_demo_01 = self.browse_ref("dms.directory_04_demo")` This folder only contains a md file and it is working fine locally. Can you try, If it works, we can try the same on 13 :thinking:

Commit:
[FIX] dms: fix wrong test

In v13, this test is programmed in such a way that the demo user is supposed to be able to copy that subdirectory: https://github.com/OCA/dms/blob/c3f802db43362127e70d8c7b4987fb71d4c1f01c/dms/tests/test_directory.py#L40

However, in https://github.com/OCA/dms/pull/7 that test was modified indicating that demo user didn't have permissions to do that: https://github.com/OCA/dms/blob/e3b6d8d24534f2a68bfb88e310cc70cefe46bb64/dms/tests/test_directory.py#L39

Rolling back that change to ensure premissions remain the same in both versions of the module.

Also changing the directory to test to ensure it contains no SVG files, whose detection seems to differ among environments, and which have some specific permission restrictions that can make the modified test fail or pass.

@Tecnativa TT25645

yajo commented:
It seems Odoo 12 can use 3 different implementation of `guess_mimetype`:

1. [The one bundled with Odoo](https://github.com/odoo/odoo/blob/d201fdd75946fa75bcce3dc0be7203649e7972db/odoo/tools/mimetypes.py#L135-L161).
1. https://pypi.python.org/pypi/python-magic/ [imported here](https://github.com/odoo/odoo/blob/d201fdd75946fa75bcce3dc0be7203649e7972db/odoo/tools/mimetypes.py#L173).
1. https://packages.debian.org/squeeze/python-magic (broken link) [imported here](https://github.com/odoo/odoo/blob/d201fdd75946fa75bcce3dc0be7203649e7972db/odoo/tools/mimetypes.py#L178).

So, maybe I have a different set of libraries installed in doodba than the ones installed in travis, and maybe that's why on one environment the file is detected as SVG and on the other one it isn't.



> Maybe we can replace `self.directory_sub_demo_01 = self.browse_ref("dms.directory_03_demo")` by `self.directory_sub_demo_01 = self.browse_ref("dms.directory_04_demo")`

I tried on doodba and it goes :heavy_check_mark:. Let's see here, I pushed the change.

etobella commented:
Great I will wait until the test of the PR on 13.0 works fine and proceed with merges :wink:

pedrobaeza commented on dms/tests/test_directory.py:
```diff
@@ -36,7 +36,7 @@ def test_copy_root_directory(self):
         )
 
-    @multi_users(lambda self: self.multi_users(demo=False), callback="_setup_test_data")
+    @multi_users(lambda self: self.multi_users(), callback="_setup_test_data")
```
Why removing this then?

etobella replied:
Originally it was not here, and It was added because an error was raised, but the error was raised due to an error to the mimetype detection. Travis was detecting that one of the files of the folder was an SVG. Now we are using a folder that does not has this kind of records and is not giving problems.
So, it was a mistake to add `demo=False`. This PR solves the issue and now it is consistent with 13.0

pedrobaeza replied:
OK

pedrobaeza commented:
@Yajo it's interesting to check if in Doodba we should provide the Python library for better mimetype detection.

pedrobaeza commented:
/ocabot merge patch

pedrobaeza commented:
The bot went crazy...
